### ISSUE TITLE
Provide generic @cache decorator

### ISSUE THREAD

**Issue by tkem** [tkem/cachetools#34]
A generic `@cachedfunc` decorator should work with *any* mutable mapping, similar to `@cachedmethod`. One issue is with `cache_info()` support for `currsize` and `maxsize` (which `dict` does not have, but this might be handled with optional "getter" parameters, e.g.
```
def cachedfunc(cache_factory, typed=False, currsize=len, maxsize=None, lock=None)
```
Note that to be generally usable as a decorator, the first argument should be a factory function, which could be used like
```
@cachedfunc(dict)
def myfun(arg1, arg2)
```
In the `@lru_cache` compatible decorators, this could be passed a lambda, e.g.
```
def lru_cache(maxsize=128, typed=False, getsizeof=None, lock=RLock):
    return cachedfunc(lambda: LRUCache(maxsize, getsizeof), 
                                      typed, 
                                      currsize=operator.attrgetter('currsize'),
                                      maxsize=operator.attrgetter('maxsize'),
                                      lock=lock)
```
Alternatively, an `info()` method could be added to the `Cache` base class, tracking cache hits and misses, but this would probably conflict with how cache implementations use the base class's `__getitem__`. For example, `TTLCache` and `LRUCache` call `Cache.__getitem__` during `__setitem__`, which would increment cache hits/misses.

On the other hand, providing `currsize` and `maxsize` for `@cachedfunc` might call for similar support in `@cachedmethod`, which has to be evaluated seperately.

**Comment by tkem** [comment]
Actually, the first argument *may* be a cache *object*, as exemplified by 0k/kids.cache - benefits of a factory function should be reevaluated.

**Comment by tkem** [comment]
As for the `currsize` and `maxsize` properties, we could probably get away with checking for the respective attributes, falling back on `len` if `cursize` is not available.

**Comment by tkem** [comment]
If the current Python 3 style decorators are moved to a sub module as suggested in #39, there's no real need for the API to be Py3 compliant any more.
So `cache_info()` etc. can be removed for the generic function decorator, and replaced with some way to access the cache object directly, e.g. through a function property.

**Comment by tkem** [comment]
Proposed interface:
```
@cache(cache, key, context)
```
with
- `cache` being a cache *object* (or `dict`, or other mapping object)
- `key` being a function that returns a hashable key for `(func, *args, **kwargs`)
- `context` being a context manager object that's invoked every time the cache object is accessed, similar to the current `lock`

The Py3 compatible decorators would use `context` for keeping track of cache access/misses to implement `CacheInfo`. This needs to be prototyped as P.O.C.

`@cachedmethod` should uyse the same interface, except for `cache` being a function.

Botyh `key` and `context` should be optional.

**Comment by tkem** [comment]
All three parameters should be available as properties of the function wrapper. The wrapped function itself should be made available as `__wrapped__` on Python 2.7.

**Comment by tkem** [comment]
Also need to prototype use cases of #36 with this interface, preferably as unit tests.
